Use static analysis tools

Static analysis tools will help improve code quality and avoid common mistakes.

Suggestions:
- [`pylint`](https://pylint.org/): code quality and common issues
- [`mypy`](http://mypy-lang.org/): static type checking
- [`pyroma`](https://github.com/regebro/pyroma#version-control-integration): check package quality and `setup.py`

In the current codebase:

`pylint`:
```text
mariadb_kernel\maria_magics\line_magic.py:21:0: W0223: Method 'execute' is abstract in class 'MariaMagic' but is not overridden (abstract-method)
mariadb_kernel\maria_magics\line_magic.py:21:0: W0223: Method 'help' is abstract in class 'MariaMagic' but is not overridden (abstract-method)
mariadb_kernel\maria_magics\line_magic.py:21:0: W0223: Method 'name' is abstract in class 'MariaMagic' but is not overridden (abstract-method)
mariadb_kernel\maria_magics\line_magic.py:25:4: W0105: String statement has no effect (pointless-string-statement)
```

`pyroma`:
```console
# local clone
$ pyroma .
------------------------------
Checking .
Found mariadb_kernel
------------------------------
The classifiers should specify what minor versions of Python you support as well as what major version.
Your package does not have keywords data.
Specifying a development status in the classifiers gives users a hint of how stable your software is.
------------------------------
Final rating: 9/10
Cottage Cheese
------------------------------

# checking PyPI
$ pyroma --pypi mariadb_kernel
------------------------------
Checking mariadb_kernel
Starting new HTTPS connection (1): pypi.org:443
https://pypi.org:443 "GET /pypi/mariadb_kernel/json HTTP/1.1" 301 222
https://pypi.org:443 "GET /pypi/mariadb-kernel/json HTTP/1.1" 200 3867
Found mariadb_kernel version 0.1.1
Downloading mariadb_kernel-0.1.1.tar.gz to verify distribution
Starting new HTTPS connection (1): files.pythonhosted.org:443
https://files.pythonhosted.org:443 "GET /packages/2f/d4/adad2d7816c6f49fccc36d054b837fb270b37df5c05b943237e44bb64cd9/mariadb_kernel-0.1.1.tar.gz HTTP/1.1" 200 4683121
Found mariadb-kernel
------------------------------
The classifiers should specify what minor versions of Python you support as well as what major version.
Your package does not have keywords data.
You should have three or more owners of the project on PyPI.
Specifying a development status in the classifiers gives users a hint of how stable your software is.
------------------------------
Final rating: 8/10
Philadelphia
------------------------------
```

--- a/.github/workflows/pylint.yml
+++ b/.github/workflows/pylint.yml
@@ -1,0 +1,24 @@
+name: pylint
+
+on:
+  push:
+  pull_request:
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  pylint:
+    runs-on: ubuntu-latest
+    container: robertbindar/mariadb_jupyter_kernel:latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install dependencies
+        run: |
+          python -m pip install -r dev-requirements.txt
+          python setup.py develop
+
+      - name: Test with pylint
+        run: |
+          pylint mariadb_kernel/*.py

--- a/.pylintrc
+++ b/.pylintrc
@@ -1,0 +1,14 @@
+[MASTER]
+ignore=_version.py
+
+[MESSAGES CONTROL]
+disable=
+    missing-docstring,
+    missing-module-docstring,
+    fixme,
+    no-name-in-module,
+    import-error,
+    broad-except,
+    too-many-arguments,
+    unused-argument,
+    abstract-method

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -1,2 +1,3 @@
 pytest
 notebook
+pylint

--- a/mariadb_kernel/client_config.py
+++ b/mariadb_kernel/client_config.py
@@ -53,17 +53,17 @@ class ClientConfig:
         using_default = False
 
         try:
-            f = open(path, "r")
-            cfg = json.load(f)
-        except (OSError, json.JSONDecodeError) as e:
-            if isinstance(e, OSError):
+            with open(path, "r", encoding="utf-8") as file:
+                cfg = json.load(file)
+        except (OSError, json.JSONDecodeError) as exception:
+            if isinstance(exception, OSError):
                 self.log.info(
                     f"Config file {self.config_name} at {path} " "does not exist"
                 )
-            if isinstance(e, json.JSONDecodeError):
+            if isinstance(exception, json.JSONDecodeError):
                 self.log.info(
                     f"Config file {self.config_name} at {path} "
-                    f"is not valid JSON: {e}"
+                    f"is not valid JSON: {exception}"
                 )
             using_default = True
 
@@ -94,34 +94,34 @@ class ClientConfig:
         return os.path.join(default_dir, self.config_name)
 
     def get_args(self):
-        rv = ""
+        params = ""
         keys = ["user", "host", "port", "password", "socket", "database"]
         for key in keys:
             value = self.default_config[key]
             if value is not None:
-                rv += f"--{key}={value} "
+                params += f"--{key}={value} "
 
         # Disable progress reports in statements like LOAD DATA
-        rv += "--disable-progress-reports"
-        return rv
+        params += "--disable-progress-reports"
+        return params
 
     def get_server_args(self):
-        rv = []
-        rv.extend(self.default_config["extra_server_config"])
+        params = []
+        params.extend(self.default_config["extra_server_config"])
         # Use same connection config for both server and client
-        rv.append(f"--socket={self.default_config['socket']}")
-        rv.append(f"--port={self.default_config['port']}")
-        rv.append(f"--bind-address={self.default_config['host']}")
+        params.append(f"--socket={self.default_config['socket']}")
+        params.append(f"--port={self.default_config['port']}")
+        params.append(f"--bind-address={self.default_config['host']}")
         # Server specific config
-        rv.append(f"--datadir={self.default_config['server_datadir']}")
-        rv.append(f"--pid-file={self.default_config['server_pid']}")
-        return rv
+        params.append(f"--datadir={self.default_config['server_datadir']}")
+        params.append(f"--pid-file={self.default_config['server_pid']}")
+        return params
 
     def get_init_args(self):
-        rv = []
-        rv.extend(self.get_server_args())
-        rv.extend(self.default_config["extra_db_init_config"])
-        return rv
+        params = []
+        params.extend(self.get_server_args())
+        params.extend(self.default_config["extra_db_init_config"])
+        return params
 
     def get_server_paths(self):
         return [

--- a/mariadb_kernel/install.py
+++ b/mariadb_kernel/install.py
@@ -14,15 +14,15 @@ kernel_json = {
 
 
 def install_my_kernel_spec(user=True, prefix=None):
-    with TemporaryDirectory() as td:
-        os.chmod(td, 0o755)  # Starts off as 700, not user readable
-        with open(os.path.join(td, "kernel.json"), "w") as f:
-            json.dump(kernel_json, f, sort_keys=True)
+    with TemporaryDirectory() as temp:
+        os.chmod(temp, 0o755)  # Starts off as 700, not user readable
+        with open(os.path.join(temp, "kernel.json"), "w", encoding="utf-8") as file:
+            json.dump(kernel_json, file, sort_keys=True)
         # TODO: Copy any resources
 
         print("Installing Jupyter kernel spec")
         KernelSpecManager().install_kernel_spec(
-            td, "mariadb_kernel", user=user, prefix=prefix
+            temp, "mariadb_kernel", user=user, prefix=prefix
         )
 
 
@@ -34,23 +34,23 @@ def _is_root():
 
 
 def main(argv=None):
-    ap = argparse.ArgumentParser()
-    ap.add_argument(
+    arg_parse = argparse.ArgumentParser()
+    arg_parse.add_argument(
         "--user",
         action="store_true",
         help="Install to the per-user kernels registry. Default if not root.",
     )
-    ap.add_argument(
+    arg_parse.add_argument(
         "--sys-prefix",
         action="store_true",
         help="Install to sys.prefix (e.g. a virtualenv or conda env)",
     )
-    ap.add_argument(
+    arg_parse.add_argument(
         "--prefix",
         help="Install to the given prefix. "
         "Kernelspec will be installed in {PREFIX}/share/jupyter/kernels/",
     )
-    args = ap.parse_args(argv)
+    args = arg_parse.parse_args(argv)
 
     if args.sys_prefix:
         args.prefix = sys.prefix

--- a/mariadb_kernel/kernel.py
+++ b/mariadb_kernel/kernel.py
@@ -2,9 +2,14 @@
 # Copyright (c) MariaDB Foundation.
 # Distributed under the terms of the Modified BSD License.
 
+import os
+import signal
+import logging
+import pandas
+
 from ipykernel.kernelbase import Kernel
 
-from ._version import version as __version__
+from mariadb_kernel._version import version as __version__
 from mariadb_kernel.client_config import ClientConfig
 from mariadb_kernel.mariadb_client import (
     MariaDBClient,
@@ -16,10 +21,6 @@ from .code_completion.sql_fetch import SqlFetch
 from .code_completion.autocompleter import Autocompleter
 from .code_completion.introspector import Introspector
 
-import logging
-import pandas
-import os
-import signal
 
 _EXPERIMENTAL_KEY_NAME = "_jupyter_types_experimental"
 
@@ -77,7 +78,7 @@ class MariaDBKernel(Kernel):
                     self.mariadb_client, self.client_config, self.log
                 )
                 self.introspector = Introspector()
-            except:
+            except Exception:
                 # Something went terribly wrong, disabling the feature
                 self.log.error(
                     "Code completion functionalities were disabled due to an unexpected error"
@@ -100,8 +101,8 @@ class MariaDBKernel(Kernel):
         if not result_html or not result_html.startswith("<TABLE"):
             return
 
-        df = pandas.read_html(result_html)
-        self.data["last_select"] = df[0]
+        data = pandas.read_html(result_html)
+        self.data["last_select"] = data[0]
 
     def _send_message(self, stream, message):
         error = {"name": stream, "text": message + "\n"}
@@ -110,7 +111,7 @@ class MariaDBKernel(Kernel):
     def do_execute(
         self, code, silent, store_history=True, user_expressions=None, allow_stdin=False
     ):
-        rv = {
+        result = {
             "status": "ok",
             # The base class increments the execution count
             "execution_count": self.execution_count,
@@ -120,13 +121,13 @@ class MariaDBKernel(Kernel):
 
         try:
             parser = CodeParser(self.log, code, self.delimiter)
-        except ValueError as e:
-            self._send_message("stderr", str(e))
-            return rv
+        except ValueError as exception:
+            self._send_message("stderr", str(exception))
+            return result
 
         statements = parser.get_sql()
-        for s in statements:
-            result = self.mariadb_client.run_statement(s)
+        for statement in statements:
+            result = self.mariadb_client.run_statement(statement)
 
             if self.mariadb_client.iserror():
                 self._send_message("stderr", self.mariadb_client.error_message())
@@ -148,19 +149,19 @@ class MariaDBKernel(Kernel):
         if self.autocompleter:
             self.autocompleter.refresh()
 
-        return rv
+        return result
 
     def kill_server(self):
         if self.mariadb_server and self.mariadb_server.is_up():
-            self.log.info(f"Stopping (own) MariaDB server")
+            self.log.info("Stopping (own) MariaDB server")
             self.mariadb_server.stop()
         elif not self.mariadb_server:
             pidfile = self.client_config.get_server_pidfile()
             try:
-                with open(pidfile, "r") as f:
-                    pid = int(f.read())
-            except Exception as e:
-                self.log.error(f"Failed reading pid file {pidfile}, error: {e}")
+                with open(pidfile, "r", encoding="utf-8") as file:
+                    pid = int(file.read())
+            except Exception as exception:
+                self.log.error(f"Failed reading pid file {pidfile}, error: {exception}")
                 return
 
             if pid:
@@ -175,9 +176,9 @@ class MariaDBKernel(Kernel):
             try:
                 sql_fetch = SqlFetch(self.mariadb_client, self.log)
                 num_clients = sql_fetch.num_connected_clients()
-            except Exception as e:
+            except Exception as exception:
                 self.log.error(
-                    f"Failed querying server (of number of clients connected), error: {e}"
+                    f"Failed querying server (of number of clients connected), error: {exception}"
                 )
 
         self.mariadb_client.stop()
@@ -206,12 +207,13 @@ class MariaDBKernel(Kernel):
         for completion in completion_list:
             if completion.display_meta is not None:
                 type_dict_list.append(
-                    dict(
-                        start=completion.start_position,
-                        end=len(completion.text) + completion.start_position,
-                        text=completion.text,
-                        type=completion.display_meta_text,  # display_meta is FormattedText object
-                    )
+                    {
+                        "start": completion.start_position,
+                        "end": len(completion.text) + completion.start_position,
+                        "text": completion.text,
+                        # display_meta is FormattedText object
+                        "type": completion.display_meta_text,
+                    }
                 )
         return {
             "status": "ok",
@@ -221,7 +223,7 @@ class MariaDBKernel(Kernel):
             "metadata": {_EXPERIMENTAL_KEY_NAME: type_dict_list},
         }
 
-    def do_inspect(self, code, cursor_pos, detail_level):
+    def do_inspect(self, code, cursor_pos, detail_level=0):
         empty_result = {"status": "ok", "data": {}, "metadata": {}, "found": False}
         if not self.introspector:
             return empty_result

--- a/mariadb_kernel/mariadb_client.py
+++ b/mariadb_kernel/mariadb_client.py
@@ -3,11 +3,11 @@
 # Copyright (c) MariaDB Foundation.
 # Distributed under the terms of the Modified BSD License.
 
-from pexpect import replwrap, EOF, TIMEOUT, ExceptionPexpect
-from pathlib import Path
 import re
+from pathlib import Path
 from uuid import uuid1
 from bs4 import BeautifulSoup
+from pexpect import replwrap, EOF, TIMEOUT, ExceptionPexpect
 
 
 class MariaREPL(replwrap.REPLWrapper):
@@ -20,7 +20,7 @@ class MariaREPL(replwrap.REPLWrapper):
         patterns = [self.prompt]
         return self.child.expect(patterns, timeout=timeout, async_=async_)
 
-    def run_command(self, code, timeout=-1, async_=False):
+    def run_command(self, command, timeout=-1, async_=False):
 
         # Writing the cell code within a file and then sourcing it in the client
         # offers us a lot of advantages.
@@ -29,12 +29,12 @@ class MariaREPL(replwrap.REPLWrapper):
         # sending continuation prompt when "\n" is received.
         stmt_file = ".mariadb_statement" + "_" + str(uuid1())
         statement_file_path = Path.cwd().joinpath(stmt_file)
-        with statement_file_path.open("w") as f:
-            f.write(code)
+        with statement_file_path.open("w", encoding="utf-8") as file:
+            file.write(command)
         self.child.sendline(f"source {str(statement_file_path)}")
 
         try:
-            pattern = self._expect_prompt(timeout, async_)
+            self._expect_prompt(timeout, async_)
         finally:
             statement_file_path.unlink()
 
@@ -72,18 +72,18 @@ class MariaDBClient:
         try:
             self._launch_client()
             self.log.info("MariaDB client was successfully started")
-        except EOF as e:
+        except EOF as exception:
             self.log.error("MariaDB client failed to start")
 
-            if "Access denied for user" in e.value:
+            if "Access denied for user" in exception.value:
                 self.log.error("The credentials used for connecting are wrong")
-                raise LoginError()
+                raise LoginError from exception
 
             self.log.error("Most probably the MariaDB server is not started")
 
             # Let the kernel know the server is down
-            raise ServerIsDownError()
-        except ExceptionPexpect as e:
+            raise ServerIsDownError from exception
+        except ExceptionPexpect:
             self.log.error(
                 "No mariadb> command line client found at " f"{self.client_bin};"
             )
@@ -106,16 +106,16 @@ class MariaDBClient:
         result = ""
         try:
             result = self.maria_repl.run_command(code, timeout)
-        except EOF as e:
+        except EOF as exception:
             self.log.error(
                 f'MariaDB client failed to run command "{code}". '
-                f"Client most probably exited due to a crash: {e}"
+                f"Client most probably exited due to a crash: {exception}"
             )
             # TODO: attempt a restart and raise exception if it fails again
-        except TIMEOUT as e:
+        except TIMEOUT as exception:
             self.log.error(
                 f'MariaDB client failed to run command "{code}". '
-                f"Reading from the client timed out: {e}"
+                f"Reading from the client timed out: {exception}"
             )
             # TODO: attempt to rerun the cmd and raise exception if failure
 
@@ -125,12 +125,13 @@ class MariaDBClient:
             # Get rid of extra info in the error message that isn't interesting
             # in this case.
             # This matches error messages that look like:
-            # """ERROR 1064 (42000) at line 1 in file: './mariadb_statement': You have an error..."""
+            # """ERROR 1064 (42000) at line 1 in file:
+            # './mariadb_statement': You have an error..."""
             # We only keep the SQL error message and discard the first part
             regex = re.compile(r"^ERROR.+in file: \'.+mariadb_statement.+\': ")
             self.errormsg = regex.sub("", result, count=1)
             return self.errormsg
-        elif not result:
+        if not result:
             result = "Query OK"
 
         self.error = False

--- a/mariadb_kernel/mariadb_server.py
+++ b/mariadb_kernel/mariadb_server.py
@@ -35,6 +35,7 @@ class MariaDBServer:
             os.makedirs(path, exist_ok=True)
         try:
             self.init_db()
+            # pylint: disable=consider-using-with # setting a class member, not using directly here
             self.server = subprocess.Popen(
                 cmd,
                 stdout=subprocess.PIPE,
@@ -68,6 +69,7 @@ class MariaDBServer:
                 cmd,
                 capture_output=True,
                 universal_newlines=True,
+                check=True,
             )
         except FileNotFoundError:
             self.log.error(f"Could not find {db_init_bin}")
@@ -76,8 +78,6 @@ class MariaDBServer:
         if init_process.returncode > 0:
             self.log.error("Init failed, output:")
             self.log.error(init_process.stderr)
-
-        return
 
     def stop(self):
         if not self.is_up():

